### PR TITLE
[SPARK-46115][SQL] Restrict charsets in `encode()`

### PR DESCRIPTION
**Pull request by MaxGekk**
### What changes were proposed in this pull request?
In the PR, I propose to restrict the supported charsets in the `encode()` functions by the list from [the doc](https://spark.apache.org/docs/latest/api/sql/#encode):
```
'US-ASCII', 'ISO-8859-1', 'UTF-8', 'UTF-16BE', 'UTF-16LE', 'UTF-16'
```
and introduce the SQL config `spark.sql.legacy.javaCharsets` for restoring the previous behaviour.

### Why are the changes needed?
Currently the list of supported charsets in `encode()` is not stable and fully depends on the used JDK version. So, sometimes user code might not work because a devop changed Java version in Spark cluster.

### Does this PR introduce _any_ user-facing change?
Yes.

### How was this patch tested?
By running new checks:
```
$ PYSPARK_PYTHON=python3 build/sbt "sql/testOnly org.apache.spark.sql.SQLQueryTestSuite -- -z string-functions.sql"
```

### Was this patch authored or co-authored using generative AI tooling?
No.

### CODE REVIEW

**Review by HyukjinKwon — APPROVED** [review]

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 1]
```diff
@@ -4584,6 +4584,15 @@ object SQLConf {
...
     .createWithDefault(1)
 
+  val LEGACY_JAVA_CHARSETS = buildConf("spark.sql.legacy.javaCharsets")
+    .internal()
+    .doc("When set to true, the functions like `encode()` can use charsets from JDK while " +
+      "encoding or decoding string values. If it is false, such functions support only one of " +
+      "the charsets: 'US-ASCII', 'ISO-8859-1', 'UTF-8', 'UTF-16BE', 'UTF-16LE', 'UTF-16'.")
+    .version("4.0.0")
+    .booleanConf
+    .createWithDefault(false)
```
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:4594]: should probably update migration guide

**Review by srielau — APPROVED** [review]

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala [hunk 2]
```diff
@@ -2685,9 +2685,12 @@ case class StringDecode(bin: Expression, charset: Expression)
   since = "1.5.0",
   group = "string_funcs")
 // scalastyle:on line.size.limit
-case class Encode(value: Expression, charset: Expression)
+case class Encode(value: Expression, charset: Expression, legacyCharsets: Boolean)
```
Comment (beliefer) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala:2688]: Do we really need the new parameter? It seems we can use `SQLConf.get.legacyJavaCharsets` directly.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala [hunk 3]
```diff
@@ -2696,7 +2699,13 @@ case class Encode(value: Expression, charset: Expression)
   protected override def nullSafeEval(input1: Any, input2: Any): Any = {
     val toCharset = input2.asInstanceOf[UTF8String].toString
     try {
-      input1.asInstanceOf[UTF8String].toString.getBytes(toCharset)
+      if (legacyCharsets ||
+        toCharset.equalsIgnoreCase("UTF-8") || toCharset.equalsIgnoreCase("US-ASCII") ||
+        toCharset.equalsIgnoreCase("ISO-8859-1") ||
+        toCharset.equalsIgnoreCase("UTF-16") || toCharset.equalsIgnoreCase("UTF-16LE") ||
+        toCharset.equalsIgnoreCase("UTF-16BE")) {
```
Comment (beliefer) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala:2706]: Shall we add `val supportedCharsets = Set("UTF-8", ...)`? So we can simplify these code.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala (hunk 1 above)
Comment (beliefer) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:4594]: +1

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala (hunk 2 above)
Comment (MaxGekk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala:2688]: This avoids some issues, see the motivation in the PR's description: https://github.com/apache/spark/pull/27658

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala (hunk 3 above)
Comment (MaxGekk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala]: Added

**Comment by HyukjinKwon** [comment]
Merged to master.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala (hunk 2 above)
Comment (beliefer) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala:2688]: Got it.

**Pull request closed without merging**